server/src/main/java/org/opensearch/action/search/CreatePitController.java

```diff
@@ -0,0 +1,268 @@
+                            groupedActionListener
+                        );
+                    } catch (Exception e) {
+                        logger.error(() -> new ParameterizedMessage("Create pit update phase failed on node [{}]", node), e);
+                        groupedActionListener.onFailure(new OpenSearchException("Create pit failed on node[" + node + "]", e));
+                    }
+                }
+            }, updatePitIdListener::onFailure);
+        }, updatePitIdListener::onFailure);
+    }
+
+    private StepListener<BiFunction<String, String, DiscoveryNode>> getConnectionLookupListener(SearchContextId contextId) {
+        ClusterState state = clusterService.state();
+
+        final Set<String> clusters = contextId.shards()
+            .values()
+            .stream()
+            .filter(ctx -> Strings.isEmpty(ctx.getClusterAlias()) == false)
+            .map(SearchContextIdForNode::getClusterAlias)
+            .collect(Collectors.toSet());
+
+        final StepListener<BiFunction<String, String, DiscoveryNode>> lookupListener = new StepListener<>();
+
+        if (clusters.isEmpty()) {
+            lookupListener.onResponse((cluster, nodeId) -> state.getNodes().get(nodeId));
+        } else {
+            searchTransportService.getRemoteClusterService().collectNodes(clusters, lookupListener);
+        }
+        return lookupListener;
+    }
+
+    private ActionListener<UpdatePitContextResponse> getGroupedListener(
+        ActionListener<CreatePitResponse> updatePitIdListener,
+        CreatePitResponse createPITResponse,
+        int size,
+        Collection<SearchContextIdForNode> contexts
+    ) {
+        return new GroupedActionListener<>(new ActionListener<>() {
+            @Override
+            public void onResponse(final Collection<UpdatePitContextResponse> responses) {
+                updatePitIdListener.onResponse(createPITResponse);
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                cleanupContexts(contexts);
+                updatePitIdListener.onFailure(e);
+            }
+        }, size);
+    }
+
+    /**
+     * Cleanup all created PIT contexts in case of failure
+     */
+    private void cleanupContexts(Collection<SearchContextIdForNode> contexts) {
+        ActionListener<Integer> deleteListener = new ActionListener<>() {
+            @Override
+            public void onResponse(Integer freed) {
+                // log the number of freed contexts - this is invoke and forget call
+                logger.debug(() -> new ParameterizedMessage("Cleaned up {} contexts out of {}", freed, contexts.size()));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error("Cleaning up PIT contexts failed ", e);
+            }
+        };
+        ClearScrollController.closeContexts(clusterService.state().getNodes(), searchTransportService, contexts, deleteListener);
```

Comment:
Can we abstract the common logic out that would aid scroll deprecation in future

Reply:
Will change this logic to use delete pit flows as part of delete PIT PR so that we don't have any scroll dependencies.